### PR TITLE
Forcibly removing empty or bad metadata files from FS for DiskS3

### PR DESCRIPTION
**Pull request by Jokser**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Forcibly removing empty or bad metadata files from filesystem for DiskS3. S3 is an experimental feature.

Detailed description / Documentation draft:
When there is no space on FS there is a chance that metadata file created for DiskS3 will be empty:

> 2020.11.27 13:41:53.050126 [ 3592 ] {} <Error> virtual DB::CompletionAwareWriteBuffer::~CompletionAwareWriteBuffer(): Code: 75, e.displayText() = DB::ErrnoException: Cannot write to file /var/lib/clickhouse/disks/object_storage/data/inflow/appsflyer/tmp_merge_fb46e33a8c1e4d548fbc321657b0b2e8_72690_72705_1/event_name.mrk2, errno: 28, strerror: No space left on device, Stack trace (when copying this message, always include the lines below):

After that such files cant be removed:
2020.11.27 13:41:54.871161 [ 3611 ] {} <Error> inflow.appsflyer: Code: 32, e.displayText() = DB::Exception: Attempt to read after eof, Stack trace (when copying this message, always include the lines below):
0. DB::throwReadAfterEOF() @ 0x7b65dcd in /usr/lib/debug/usr/bin/clickhouse
1. void DB::readIntTextImpl<unsigned int, void, (DB::ReadIntTextCheckOverflow)0>(unsigned int&, DB::ReadBuffer&) @ 0x7c94c9f in /usr/lib/debug/usr/bin/clickhouse
2. DB::(anonymous namespace)::Metadata::Metadata(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) @ 0xd82bc9f in /usr/lib/debug/usr/bin/clickhouse
3. DB::DiskS3::removeMeta(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::DiskS3::AwsS3KeyKeeper&) @ 0xd8323fd in /usr/lib/debug/usr/bin/clickhouse
4. DB::DiskS3::removeMetaRecursive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::DiskS3::AwsS3KeyKeeper&) @ 0xd83290c in /usr/lib/debug/usr/bin/clickhouse
5. DB::DiskS3::removeMetaRecursive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::DiskS3::AwsS3KeyKeeper&) @ 0xd832962 in /usr/lib/debug/usr/bin/clickhouse
6. DB::DiskS3::removeRecursive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xd834bc5 in /usr/lib/debug/usr/bin/clickhouse
7. DB::IMergeTreeDataPart::makeCloneOnDisk(std::__1::shared_ptr<DB::IDisk> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0xe2bb750 in /usr/lib/debug/usr/bin/clickhouse
8. DB::MergeTreePartsMover::clonePart(DB::MergeTreeMoveEntry const&) const @ 0xe3e5318 in /usr/lib/debug/usr/bin/clickhouse
9. DB::MergeTreeData::moveParts(DB::MergeTreeData::CurrentlyMovingPartsTagger&&) @ 0xe317fb9 in /usr/lib/debug/usr/bin/clickhouse
10. DB::StorageMergeTree::movePartsTask() @ 0xe0d95b9 in /usr/lib/debug/usr/bin/clickhouse
11. DB::BackgroundProcessingPool::workLoopFunc() @ 0xe297663 in /usr/lib/debug/usr/bin/clickhouse
12. std::__1::__function::__func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundProcessingPool::BackgroundProcessingPool(int, DB::BackgroundProcessingPool::PoolSettings const&, char const*, char const*)::$_0>(DB::BackgroundProcessingPool::BackgroundProcessingPool(int, DB::BackgroundProcessingPool::PoolSettings const&, char const*, char const*)::$_0&&)::'lambda'(), std::__1::allocator<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundProcessingPool::BackgroundProcessingPool(int, DB::BackgroundProcessingPool::PoolSettings const&, char const*, char const*)::$_0>(DB::BackgroundProcessingPool::BackgroundProcessingPool(int, DB::BackgroundProcessingPool::PoolSettings const&, char const*, char const*)::$_0&&)::'lambda'()>, void ()>::operator()() @ 0xe298193 in /usr/lib/debug/usr/bin/clickhouse
13. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x7b8b63d in /usr/lib/debug/usr/bin/clickhouse
14. void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0x7b8f153 in /usr/lib/debug/usr/bin/clickhouse
15. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
16. clone @ 0x121a3f in /lib/x86_64-linux-gnu/libc-2.27.so
 (version 20.10.4.1 (official build))

We should make it possible to remove such metadata files from FS forcibly to not break ClickHouse.

### CODE REVIEW

File: src/Disks/S3/DiskS3.cpp [hunk 1]
```diff
@@ -703,9 +714,11 @@ void DiskS3::removeMeta(const String & path, AwsS3KeyKeeper & keys)
             file.remove();
         }
     }
-    else
+    catch (...)
```
Comment (alexey-milovidov) [inline src/Disks/S3/DiskS3.cpp:717]: This is dangerous, e.g. "memory limit exceeded".

Comment (alexey-milovidov) [reply src/Disks/S3/DiskS3.cpp:717]: Maybe check for some set of error codes explicitly?

File: src/Disks/S3/DiskS3.cpp [hunk 2]
```diff
@@ -63,332 +63,339 @@ void DiskS3::AwsS3KeyKeeper::addKey(const String & key)
...
+        total_size += size;
+        s3_objects.emplace_back(path, size);
+    }
 
-            writeIntText(VERSION_RELATIVE_PATHS, buf);
-            writeChar('\n', buf);
+    /// Fsync metadata file if 'sync' flag is set.
+    void save(bool sync = false)
+    {
+        WriteBufferFromFile buf(disk_path + metadata_file_path, 1024);
```
Comment (alexey-milovidov) [inline src/Disks/S3/DiskS3.cpp:180]: We should do:
- write to tmp file;
- fsync if need;
- rename;

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Minor changes needed.

File: src/Disks/S3/DiskS3.cpp (hunk 2 above)
Comment (Jokser) [reply src/Disks/S3/DiskS3.cpp:190]: Metadata files are already written through tmp files or directories as DiskS3 completely repeats FS layout as in the case of DiskLocal.
The main issue is to correctly remove them from FS if they can't be parsed.

File: src/Disks/S3/DiskS3.cpp (hunk 1 above)
Comment (Jokser) [reply src/Disks/S3/DiskS3.cpp]: Fixed

**Comment by alexey-milovidov** [comment]
`test_log_family_s3/test.py::test_log_family_s3[StripeLog-1-2]`

There is the time to deal with this test.

File: src/Disks/S3/DiskS3.cpp [hunk 3]
```diff
@@ -119,52 +119,62 @@ struct DiskS3::Metadata
...
+            }
+        }
+        catch (Exception & e)
         {
-            readBoolText(read_only, buf);
-            assertChar('\n', buf);
+            if (e.code() == ErrorCodes::UNKNOWN_FORMAT)
+                throw;
+
+            throw Exception("Failed to read metadata file: " + e.message(), ErrorCodes::UNKNOWN_FORMAT);
```
Comment (alexey-milovidov) [inline src/Disks/S3/DiskS3.cpp:178]: e.addMessage and throw.

Otherwise the original stack trace is lost.

File: src/Disks/S3/DiskS3.cpp [hunk 4]
```diff
@@ -714,10 +724,13 @@ void DiskS3::removeMeta(const String & path, AwsS3KeyKeeper & keys)
             file.remove();
         }
     }
-    catch (...)
+    catch (Exception & e)
```
Comment (alexey-milovidov) [inline src/Disks/S3/DiskS3.cpp:732]: Here we can add `const`.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
std::terminate in `DB::TablesBlockSource::generate()` - unrelated issue.

**Pull request merged**